### PR TITLE
[docs] Add dbt defer and Dagster branch deployments guide

### PR DESCRIPTION
**Pull request by maximearmstrong**
## Summary & Motivation

As the title. Fixes [DS-402](https://linear.app/dagster-labs/issue/DS-402/add-dbt-and-dagster-branch-deployment-guide)

## How I Tested These Changes

Docs preview

## Changelog

[dagster-dbt] A guide on how to use dbt defer with Dagster branch deployments has been added to the dbt reference.

- [ ] `NEW` _(added new feature or capability)_
- [ ] `BUGFIX` _(fixed a bug)_
- [x] `DOCS` _(added or updated documentation)_

### CODE REVIEW

**Comment by maximearmstrong** [comment]
* **#24760** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24760?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @maximearmstrong and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: docs/content/integrations/dbt/reference.mdx
Comment (slopp) [inline docs/content/integrations/dbt/reference.mdx:207]: I think this leaves a bit too much to the reader. 

Maybe we could describe how they could go about managing these artifacts and what specifically they are? Or should we suggest `dagster-cloud ci dagster-dbt project manage-state` ?

**Review by slopp — COMMENTED** [review]
Three fairly major questions:

1. How should users manage the state artifact?  Are we ready to document/suggest the `dagster-cloud` management utils?

2. I've seen users struggle with what "target" they should use in branch deployments. If they target prod (the approach used with dbt cloud slim ci) they end up in a weird state where the branch deployment has the asset materializations for the prod assets! Alternatively, if they run against a "stage" target, they need someway to re-do the dbt defer execution in prod after merging. It would be nice if we could provide some guidance here.

3. Finally, in hooli, we use a CICD step to run a special dbt job with this defer command. In this guide, it is less clear how a user should actually execute the "dynamically deferred" dbt_assets....do they need to manually run the dbt assets in the branch deployment? We should document that step.

File: docs/content/integrations/dbt/reference.mdx
Comment (maximearmstrong) [reply docs/content/integrations/dbt/reference.mdx:196]: After re-reading the code of `dagster-cloud ci dagster-dbt project manage-state`, we won't be needing the env var. Will update the guide accordingly.

File: docs/content/integrations/dbt/reference.mdx
Comment (maximearmstrong) [reply docs/content/integrations/dbt/reference.mdx]: That's fair. I updated the guide in [ebdcd56](https://github.com/dagster-io/dagster/pull/24760/commits/ebdcd562fec4065cc16cdca26d409b6a604cae42)

**Comment by maximearmstrong** [comment]
@slopp thanks for the review!

> How should users manage the state artifact? Are we ready to document/suggest the dagster-cloud management utils?

This is now mentioned after the changes made in [ebdcd56](https://github.com/dagster-io/dagster/pull/24760/commits/ebdcd562fec4065cc16cdca26d409b6a604cae42)

> I've seen users struggle with what "target" they should use in branch deployments. If they target prod (the approach used with dbt cloud slim ci) they end up in a weird state where the branch deployment has the asset materializations for the prod assets! Alternatively, if they run against a "stage" target, they need someway to re-do the dbt defer execution in prod after merging. It would be nice if we could provide some guidance here.

I think this will require some work on our end before writing a guide. 

Currently, when leveraging the `dagster-cloud` management utils, the deferred artifact will always be the production `manifest.json` file. Users should use the `staging` target in a branch deployment, so they would compare the `staging` manifest with a `prod` manifest. I'm not sure if this is the behavior we want here. 

Also, when pushing to prod, we upload the new manifest, but I think we should download the previous manifest, and keep it in the state dir. One problem with that is if two branches are push sequentially, we don't want to lose the first state before the materialization of the assets is completed.

I'll open a ticket.

> Finally, in hooli, we use a CICD step to run a special dbt job with this defer command. In this guide, it is less clear how a user should actually execute the "dynamically deferred" dbt_assets....do they need to manually run the dbt assets in the branch deployment? We should document that step.

We don't have the documentation about how to run a job using `dagster-cloud-action` in the CI/CD. I think we should have that guide first, then mention here that it is an option. I can open a ticket for that.

For now I can mention that the assets can be materialized manually. Would that make sense?

File: docs/content/integrations/dbt/reference.mdx
Comment (slopp) [inline docs/content/integrations/dbt/reference.mdx:203]: Does this command store the prod manifest at the path `state` used earlier in the code example?

Comment (maximearmstrong) [reply docs/content/integrations/dbt/reference.mdx:203]: Yes. It is mentioned below, but I can rephrase if you think it should be clearer.

File: docs/content/integrations/dbt/reference.mdx
Comment (PedramNavid) [inline docs/content/integrations/dbt/reference.mdx:164]: > This allows you to run a subset of models or tests based on the changes made since the state passed to `state_path`.

I'm not sure I fully understand this. Based on what changes?

File: docs/content/integrations/dbt/reference.mdx
Comment (PedramNavid) [inline docs/content/integrations/dbt/reference.mdx:166]:
```suggestion
In practice, this is most useful when combined with branch deployments in Dagster+ to test changes made in your branches. This can be done by updating your Dagster code and your CI/CD files.
```

Avoid 'simply, easily' etc. Nothing is simple or easy.

File: docs/content/integrations/dbt/reference.mdx
Comment (PedramNavid) [inline docs/content/integrations/dbt/reference.mdx:168]: I struggle here. You are asking me to pass a `state_path` but I don't know what `state_path` I should be passing. Where is the state saved? How is it saved? Who is creating it?

File: docs/content/integrations/dbt/reference.mdx
Comment (PedramNavid) [inline docs/content/integrations/dbt/reference.mdx:194]: Not clear to me where or when the artifacts are created

File: docs/content/integrations/dbt/reference.mdx
Comment (PedramNavid) [reply docs/content/integrations/dbt/reference.mdx:203]: I think we should lead with this and then do the code change next. It would help to have an overview section that talks through the steps first briefly. For example:


In order to be able to use `dbt defer` we will do the following:
1. Add a command to our CI/CD pipeline that saves a state file to a given location
2. Update our Dagster dbt code to reference this new state file. 
3. ...Some other step?

File: docs/content/integrations/dbt/reference.mdx
Comment (PedramNavid) [inline docs/content/integrations/dbt/reference.mdx:206]:
```suggestion
The `dagster-cloud ci dagster-dbt project manage-state` CLI command fetches the `manifest.json` file form your production branch and saves it to the state directory provided, in order to power the `dbt defer` command.
```

I would combine the two paragraphs and keep things simpler.

**Comment by slopp** [comment]
@maximearmstrong what do we need to get this across the line? Are you doing another round to address Pedram's feedback? Do you need help on that one?

File: docs/content/integrations/dbt/reference.mdx
Comment (maximearmstrong) [reply docs/content/integrations/dbt/reference.mdx]: That's fair - I've updated the content to address all your comments. All changes made in [3971f71](https://github.com/dagster-io/dagster/pull/24760/commits/3971f7182585ff67c818ff0619bba613ffa35b13)

**Comment by maximearmstrong** [comment]
> @maximearmstrong what do we need to get this across the line? Are you doing another round to address Pedram's feedback? Do you need help on that one?

@slopp just pushed another run of changes to address Pedram's comments. I think we'll be ready to merge after another run of reviews.

**Review by slopp — APPROVED** [review]
Thank you for putting this all together!

I think we should land this, and then down the road we could create an additional guide similar the "dbt + serverless" or "dbt + hybrid" documents

But this lgtm as a starting reference point

**Pull request merged**
